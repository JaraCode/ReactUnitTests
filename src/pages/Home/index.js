import React, { Component } from 'react'
import Grid from '@material-ui/core/Grid'
import { withStyles } from '@material-ui/core/styles'
import { compose } from 'redux'
import { connect } from 'react-redux'
import get from 'lodash/get'

import { getProfileData, getProfileRepos } from '../../redux/actions/profile'
import { styles } from './styles'
import RightContainer from './RightContainer'
import LeftContainer from './LeftContainer'
import { getData } from '../../utils/storage'
import { GITHUB_TOKEN } from '../../consts'

class Home extends Component {

  constructor(props) {
    super(props)
    this.state = { githubToken: null }
  }

  async componentDidMount() {
    const githubToken = getData(GITHUB_TOKEN)
    await this.setState({ githubToken })
    this.props.getProfileData({ githubToken });
  }

  componentDidUpdate(prevProps, prevState) {
    const { githubData, getProfileRepos } = this.props
    const reposUrl = get(githubData, 'repos_url', null)
    const prevReposUrl = get(prevProps, 'githubData.repos_url', null)
    if (reposUrl && (prevReposUrl !== reposUrl)) {
      getProfileRepos({ reposUrl })
    }
  }

  render() { 
    const { classes, githubData, githubUserRepos } = this.props;
    return (
      <Grid container className={classes.homeContainer}>
        <Grid item xs={3} className={classes.leftContainer}>
          <LeftContainer {...githubData} />
        </Grid>
        <Grid item xs={9} className={classes.rightContainer}>
          <RightContainer repos={githubUserRepos} />
        </Grid>
      </Grid>
    )
  }
}

const actions = {
  getProfileData,
  getProfileRepos
}

const mapStateToProps = (state) => {
  console.log(state)
  return {
    // githubData: state.githubData
    githubData: get(state, 'profile.githubData', null),
    githubUserRepos: get(state, 'profile.userRepos', null)
  }
}

export default compose(
  connect(mapStateToProps, actions),
  withStyles(styles)
)(Home)
