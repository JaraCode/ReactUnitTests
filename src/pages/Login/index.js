import React, { Component } from 'react';
import { withStyles } from '@material-ui/core/styles'
import Typography from '@material-ui/core/Typography'
import Card from '@material-ui/core/Card'
import GitHubLogin from 'react-github-login';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom'

import { getGithubToken } from '../../redux/actions/login'
import { styles } from './styles'

class Login extends Component {
  onSuccess = response => {
    if (response && response.code) {
      this.props.getGithubToken({code: response.code})
    }
  }

  onFailure = response => {
    alert("No ha sido exitoso, intenta m√°s tarde.");
  }

  render() {
    const { classes, githubToken } = this.props;

    if (githubToken) {
      return <Redirect to="/home" />
    }

    return (
      <div className={classes.container}>
        <Card className={classes.card}>
          <Typography variant="h1" className={classes.title}>Iniciar Sesion</Typography>
          <GitHubLogin
            className={classes.button}
            clientId={process.env.REACT_APP_GITHUB_CLIENT_ID}
            redirectUri={process.env.REACT_APP_GITHUB_CALLBACK_URL}
            onSuccess={this.onSuccess}
            onFailure={this.onFailure}
            buttonText="Github Account" />
        </Card>
      </div>
    )
  }
}

// mapDispatchToProps
// actions is new alternative used
const actions = {
  getGithubToken
}

const mapStateToProps = (state) => {
  console.log(state);
  return ({
  githubToken: state.login.githubToken
})}

// export default withStyles(styles)(Login);
// usually only need connect, but how this already have a HoC (withStyles),
// With Compose, can chain or merge multiple HoC to component
export default compose(
  withStyles(styles),
  connect(mapStateToProps, actions)
)(Login);

export {Login};