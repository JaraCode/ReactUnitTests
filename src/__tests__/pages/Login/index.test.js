import React from 'react';
import { shallow } from 'enzyme';

import { Login } from '../../../pages/Login';
import { Card } from '@material-ui/core';
import TestRenderer from 'react-test-renderer';

let wrapper;

describe("<Login/>", () => {


  beforeEach(() => {
    const props = {
      classes: {},
      getGithubToken: jest.fn()
    }

    wrapper = shallow(<Login {...props} />);
  })

  it('render succesful', () => {
    expect(wrapper).toBeDefined();
    expect(wrapper).toBeTruthy();
  });


  it('render a card', () => {
    // have only 1 CardComponent
    expect(wrapper.find(Card)).toHaveLength(1)
  });

  it('render properties well', () => {
    // statefull components have a instance
    // console.log(wrapper.instance())

    // stateless components dont have instance
    expect(wrapper.instance().props.classes).toBeDefined()
    expect(wrapper.instance().props.classes).toEqual({})
    expect(wrapper.instance().props.getGithubToken).toBeDefined()
  });

  it('onSuccess test', () => {
    const response = { code: 1234 }

    // is like this.onSuccess(response)
    wrapper.instance().onSuccess(response);

    expect(wrapper.instance().props.getGithubToken).toHaveBeenCalled();
  })

  it('onSuccess test with valid response', () => {
    const response = { code: 1234 }

    // is like this.onSuccess(response)
    wrapper.instance().onSuccess(response);

    expect(wrapper.instance().props.getGithubToken).toHaveBeenCalled();
  })

  it('onSuccess test with invalid response', () => {
    const response = null

    // is like this.onSuccess(response)
    wrapper.instance().onSuccess(response);

    // when has been created in beforeAll, this test fail, because other test call this.
    expect(wrapper.instance().props.getGithubToken).toHaveBeenCalledTimes(0);
  })

  it('Show alert, on wrong github response', () => {
    window.alert = jest.fn()

    // is like this.onFailure()
    wrapper.instance().onFailure({ error: 'Fail'});

    expect(window.alert).toHaveBeenCalled();
    expect(window.alert).toHaveBeenCalledTimes(1);
  })

  it("El snapshot es correcto", () => {
    const props = {
      classes: {container: "container"},
      getGithubToken: jest.fn()
    };

    const tree = TestRenderer.create(<Login {...props} />);
    // console.log(tree.toJSON());
    expect(tree.toJSON()).toMatchSnapshot();
  });

})
