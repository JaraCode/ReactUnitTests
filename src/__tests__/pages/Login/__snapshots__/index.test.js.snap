// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login/> El snapshot es correcto 1`] = `
<div
  className="container"
>
  <div
    className="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1"
  >
    <h1
      className="MuiTypography-root-29 MuiTypography-h1-41"
    >
      Iniciar Sesion
    </h1>
    <button
      onClick={[Function]}
    >
      Github Account
    </button>
  </div>
</div>
`;
